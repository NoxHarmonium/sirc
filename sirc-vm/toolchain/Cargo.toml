[package]
name = "toolchain"
version = "0.1.0"
edition = "2021"

[dependencies]
clap = { version = "4.3.19", features = ["derive"] }
peripheral_cpu = { path = "../peripheral-cpu" }
nom = "7.1.3"
nom-supreme = "0.8.0"
itertools = "0.11.0"
postcard = { version = "1.0.6", features = ["alloc"] }
serde = "1.0.181"
bytemuck = "1.13.1"
