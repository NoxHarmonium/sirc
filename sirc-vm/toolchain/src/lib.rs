#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
#![allow(
    // I don't like this rule
    clippy::module_name_repetitions,
    // Not sure what this is, will have to revisit
    clippy::must_use_candidate,
    // Will tackle this at the next clean up
    clippy::too_many_lines,
    // Might be good practice but too much work for now
    clippy::missing_errors_doc,
    // Not stable yet - try again later
    clippy::missing_const_for_fn,
    // I have a lot of temporary panics for debugging that will probably be cleaned up
    clippy::missing_panics_doc
)]
#![deny(warnings)]

pub mod data;
pub mod parsers;
pub mod types;
