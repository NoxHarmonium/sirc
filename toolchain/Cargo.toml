[package]
name = "toolchain"
version = "0.1.0"
edition = "2021"

[dependencies]
clap = { version = "3.2.22", features = ["derive"] }
peripheral_cpu = { path = "../peripheral-cpu" }
nom = "7.1.1"
itertools = "0.10.5"
postcard = { version = "1.0.2", features = ["alloc"] }
serde = "1.0.145"

[[bin]]
name = "assembler"
path = "src/assembler.rs"

[[bin]]
name = "linker"
path = "src/linker.rs"
