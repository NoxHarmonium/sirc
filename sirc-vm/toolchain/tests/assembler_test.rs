use nom_supreme::{
    error::ErrorTree,
    final_parser::{final_parser, Location},
};
use pretty_hex::{config_hex, HexConfig};
use toolchain::{
    data::object::build_object,
    parsers::instruction::{parse_tokens, Token},
};

#[test]
fn test_parse_tokens() {
    let hex_config = HexConfig {
        width: 4,
        ..HexConfig::default()
    };

    let tokens =
        match final_parser::<&str, Vec<Token>, ErrorTree<&str>, ErrorTree<Location>>(parse_tokens)(
            r#"
    ADDI r1, #1
    ADCI r1, #1
    SUBI r1, #1
    SBCI r1, #1
    ANDI r1, #1
    ORRI r1, #1
    XORI r1, #1
    LOAD r1, #1
    CMPI r1, #1
    TSAI r1, #1
    TSXI r1, #1
    COPI r1, #1
"#
            .trim_start(),
        ) {
            Ok(tokens) => tokens,
            Err(error) => panic!("Error parsing file:\n{error}"),
        };

    let object = build_object(tokens);
    insta::assert_display_snapshot!(config_hex(&object.program.as_slice(), hex_config));
}
