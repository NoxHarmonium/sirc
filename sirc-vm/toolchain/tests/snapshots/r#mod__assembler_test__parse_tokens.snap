---
source: toolchain/tests/assembler_test.rs
expression: "config_hex(&object.program.as_slice(), hex_config)"
---
Length: 48 (0x30) bytes
0000:   00 40 00 50   .@.P
0004:   04 40 00 50   .@.P
0008:   08 40 00 50   .@.P
000c:   0c 40 00 50   .@.P
0010:   10 40 00 50   .@.P
0014:   14 40 00 50   .@.P
0018:   18 40 00 50   .@.P
001c:   1c 40 00 40   .@.@
0020:   28 40 00 50   (@.P
0024:   30 40 00 50   0@.P
0028:   38 40 00 50   8@.P
002c:   3c 40 00 50   <@.P
